src/hats.ts

```diff
@@ -211,18 +228,97 @@ export function handleHatToggleChanged(event: HatToggleChanged): void {
 
 export function handleTransferSingle(event: TransferSingle): void {
   let hat = Hat.load(hatIdToHex(event.params.id)) as Hat;
-  if (event.params.from != Address.zero() && event.params.to != Address.zero()) { //transfer event
+  if (
+    event.params.from != Address.zero() &&
+    event.params.to != Address.zero()
+  ) {
+    //transfer event
     giveHat(hat, event);
     removeHat(hat, event);
-  }
-  else if (event.params.from == Address.zero() && event.params.to != Address.zero()) { // mint event
+  } else if (
+    event.params.from == Address.zero() &&
+    event.params.to != Address.zero()
+  ) {
+    // mint event
     giveHat(hat, event);
-  }
-  else if (event.params.from != Address.zero() && event.params.to == Address.zero()) { // burn event
+  } else if (
+    event.params.from != Address.zero() &&
+    event.params.to == Address.zero()
+  ) {
+    // burn event
     removeHat(hat, event);
   }
 }
 
+export function handleWearerStandingChanged(
+  event: WearerStandingChanged
+): void {
+  // load hat and update the badStandings array
+  let hat = Hat.load(hatIdToHex(event.params.hatId)) as Hat;
+  let wearer = Wearer.load(event.params.wearer.toHexString()) as Wearer;
+  let badStandings = hat.badStandings;
+  if (event.params.wearerStanding) {
+    let index = badStandings.indexOf(wearer.id);
+    badStandings.splice(index, 1);
+    hat.badStandings = badStandings;
+  } else {
+    badStandings.push(wearer.id);
+    hat.badStandings = badStandings;
+  }
+  hat.save();
+
+  // create new WearerStandingChangedEvent
+  let wearerStandingChangedEvent = new WearerStandingChangedEvent(
+    createEventID(event, "WearerStandingChanged")
+  );
+  wearerStandingChangedEvent.tree = topHatDomain(event.params.hatId);
+  wearerStandingChangedEvent.hat = hatIdToHex(event.params.hatId);
+  wearerStandingChangedEvent.blockNumber = event.block.number.toI32();
+  wearerStandingChangedEvent.timestamp = event.block.timestamp;
+  wearerStandingChangedEvent.transactionID = event.transaction.hash;
+  wearerStandingChangedEvent.wearer = wearer.id;
+  wearerStandingChangedEvent.wearerStanding = event.params.wearerStanding;
+  wearerStandingChangedEvent.save();
+}
+
+export function handleTopHatLinkRequested(event: TopHatLinkRequested): void {
+  // create new TopHatLinkRequestedEvent
+  let topHatLinkRequestedEvent = new TopHatLinkRequestedEvent(
+    createEventID(event, "TopHatLinkRequested")
+  );
+  topHatLinkRequestedEvent.tree = topHatDomainToHex(event.params.domain);
+  topHatLinkRequestedEvent.hat = topHatDomainToHatId(event.params.domain);
+  topHatLinkRequestedEvent.blockNumber = event.block.number.toI32();
+  topHatLinkRequestedEvent.timestamp = event.block.timestamp;
+  topHatLinkRequestedEvent.transactionID = event.transaction.hash;
+  topHatLinkRequestedEvent.newAdmin = hatIdToHex(event.params.newAdmin);
+  topHatLinkRequestedEvent.save();
+}
+
+export function handleTopHatLinked(event: TopHatLinked): void {
```

Comment:
Would it make sense here to update `hat.admin` and `hat.level` for the newly linked tophat? If it were costless to do so, my sense is Yes since that would bring those datapoints in sync with the protocol state.

Reply:
Similarly, but also very differently, how do you think we should handle `hat.level` for all the hats within the newly linked tree (and any trees linked to it)? 

Technically, their levels are now all changed, since `Hats.getHatLevel()` recurses through linked trees to derive the hat's level. But I'm guessing it would be quite expensive to update the data for all those hats.

Maybe some kind of lazy updating, such as if/when any of those hats are updated in some other way?